full spec for role arg spec (#80658)
ansible/ansible: docs/docsite/rst/playbook_guide/playbooks_reuse_roles.rst

    * This should be ``main`` in the case of an unspecified entry point.
    * This will be the base name of the tasks file to execute, with no ``.yml`` or ``.yaml`` file extension.

.. list-table:: Role Argument Specification Fields
  :width: 90%
  :widths: auto
  :header-rows: 1

  * - Name
    - Description
    - Type
    - Required
    - Default
  * - short_description
    - A short, one-line description of the entry point.
    - str
    - Yes
    -  
  * - description
    - A longer description that may contain multiple lines.
    - list\n elements: str
    - No (see notes)
    - null
  * - author
    - Name of the entry point authors.
    - list\n elements: str
    - No
    - []
  * - options
    - Options are often called "parameters" or "arguments".
    - dict (see table below)
    - No

>     - {}

Reviewer: ```suggestion
    - {}
  * - version_added
    - For roles in a collection, the version of the collection the entry point was added in.
    - str
    - No
    -
```

Options themselves have a defined structure

.. list-table:: Role Argument Specification options
  :width: 90%
  :widths: auto
  :header-rows: 1

  * - <Name> (the option itself)
    - Description
    - Type
    - Required
    - Default
  * - name
    - The name of the option/argument.
    - str
    - Yes
    -   
  * - description
    - Detailed explanation of what this option does. It should be written in full sentences.

>     - list\n elements: str

Reviewer: Should it be mentioned that this can also be a single string instead of a list of strings?

Author: it can, not sure if it should

    - No
    - null
  * - type
    - The data type of the option. See :ref:`Argument spec <argument_spec>` for allowed values for ``type``.
    - No
    - str
    - 
  * - required
    - Denote if the option is required or not
    - bool
    - No
    - false
  * - default
    - If an option is not required, the value is set to this if not otherwise provided
    - any (either null or should match Type above)
    - No
    - null
  * - choices
    - List of option values.
    - list
    - No
    - []
  * - elements
    - Specifies the data type for list elements when type is ``list``.
    - str
    - No
    - null
  * - options
    - If this option takes a dict or list of dicts, you can define the structure here.
    - dict (recursively use this same table)
    - No

>     - null

Reviewer: ```suggestion
    - null
  * - version_added
    - For roles in collections, the version of the collection this option was added in.
      Should only be specified if this is different from the ``version_added`` value of the entrypoint itself.
    - str
    - No
    -
```

Author: need to x2 check the code, but there does not seem to be any handling in core for this

Reviewer: It does work, `ansible-doc -t role felixfontein.acme.acme_certificate` shows `added in: version 0.6.0` for `acme_certificate_dns_substitution`.

Author: that is expected as `ansible-doc` is permissive and will not filter out what it does not know, i was checking the spec validator, but it does the same, it ignores unknown

Reviewer: Ansible-doc explicitly handles `version_added`, otherwise it would have said `version_added: 0.6.0` instead of `added in: version 0.6.0`.

I also wanted to write that there is a check for unknown parts in the argspec, but then I checked and noticed that there is a PR for it, but it never got merged: https://github.com/ansible/ansible/pull/79657

That PR knows about `version_added` (https://github.com/ansible/ansible/pull/79657/files#diff-39509a92d90332fed27ba7a3a5903ea0675dc9de161d415ccc9bf315f1dca5f6R126), and it also forces `description` to be a list of strings (https://github.com/ansible/ansible/pull/79657/files#diff-39509a92d90332fed27ba7a3a5903ea0675dc9de161d415ccc9bf315f1dca5f6R118).

  * - required
    - Denote if the option is required or not
    - bool
    - No
    - false
  * - default
    - If an option is not required, the value is set to this if not otherwise provided
    - any (either null or should match Type above)
    - No
    - null
  * - choices
    - List of option values.
    - list
    - No
    - []
  * - elements
    - Specifies the data type for list elements when type is ``list``.
    - str
    - No
    - null
  * - options
    - If this option takes a dict or list of dicts, you can define the structure here.
    - dict (recursively use this same table)
    - No
    - null
  * - version_added
    - For roles in collections, the version of the collection this option was added in.
      Should only be specified if this is different from the ``version_added`` value of the entrypoint itself.
    - str
    - No
    -
  * - version_added
    - For roles in collections, the version of the collection this option was added in.
      Should only be specified if this is different from the ``version_added`` value of the entrypoint itself.
    - str
    - No
    -

.. note::
        * The ``short_description`` is displayed by ``ansible-doc -t role -l``.

>         * description is not required by Ansible itself but antsibull-docs does expect it to be there to build documentation.

Reviewer 2: I don't know if this is a reasonable requirement to put people.
At this time there are multiple instances where ansible code itself isn't compliant[0] 

[0]https://github.com/search?q=description%3A++path%3A**%2Fargument_specs.yml+owner%3Aansible&type=code&ref=advsearch

Reviewer 2: Also, if at all possible, the type should allow for a description consisting of a string. Multiline strings beat lists in both readability and ease of writing.

Author: The ansible code is compliant, as a single string is considered a list of one element. A multiline string is still a subset of a list of strings.

Reviewer 2: No it isn't compliant, because many of those file don't have `description` key at all. 
For example:
https://github.com/ansible/instruqt/blob/a510499256cdafcb23e8a38c47ffd7343396ab21/images/ansible/roles/vs_code/meta/argument_specs.yml#L4

>         * description is not required by Ansible itself but antsibull-docs does expect it to be there to build documentation.

Reviewer: ```suggestion
        * ``description`` is not required by Ansible itself but antsibull-docs does expect it to be there to build documentation.
```

        * If a type is ``list``, ``elements`` should be specified.
        * Ensure that the default value in the docs matches the default value in the code. The actual
          default for the role variable will always come from ``defaults/main.yml``.
        * The default field should not be listed as part of the description, unless it requires additional information or conditions.

>         * If the option's type is boolean, you should use ``true``/``false`` for defaults if you want to be compatible with `ansible-lint`.

Reviewer: ```suggestion
        * If the option's type is boolean, you should use ``true``/``false`` for defaults if you want to be compatible with ``ansible-lint``.
```

>         * The options in options is recursive, so it has the same structure.

Reviewer: ```suggestion
        * The ``options`` in options is recursive, so it has the same structure.
```

    * This should be ``main`` in the case of an unspecified entry point.
    * This will be the base name of the tasks file to execute, with no ``.yml`` or ``.yaml`` file extension.

> .. csv-table:: Role Argument Specification Fields :rst:dir:`csv-table`

Reviewer 3: Looks like we mostly use rst tables
```
.. table::
   :class: documentation-table
```

However I kind of like the current documentation format. I'd be in favor of just adding the notes portion.

Author: i dont see a good way to add the diff columns (type/required/default) to the current format

Reviewer 3: Ah, yeah. Module argspec has the same issue https://docs.ansible.com/ansible/latest/dev_guide/developing_program_flow_modules.html#argument-spec. We just document the type/default if there is one in the description (none of the fields are required that I can think of). I can deal with my table-aversion, I just don't like the raw data/editing them.

    options:, 'Options are often called "parameters" or "arguments".', "dict", "No", {}

Options themselves have a defined structure:
.. csv-table:: Role Argument Specification options:rst:dir:`csv-table`

    :header: "Name", "Description", "Type", "Required", "Default"

    "name", "The name of the option/argument.", "String", "Yes", ""
    "description", "Detailed explanation of what this option does. It should be written in full sentences.", "str", "No", "null"
    "type", "The data type of the option. See :ref:`Argument spec <argument_spec>` for allowed values for ``type``.", "No",  "str"
    "required", "Denote if the option is required or not", "bool", "No", "False'
    "default", "If an option is not required, the value is set to this if not otherwise provided", "No", "null"
    "choices", "List of option values.", "No", []
    "elements", Specifies the data type for list elements when type is ``list``., "No", "null"
    "options", "If this option takes a dict or list of dicts, you can define the structure here.", "No", "null"

.. note::
        * The ``short_description`` is displayed by ``ansible-doc -t role -l``.

>         * description is not required by Ansible itself but ansibullbot does expect it to be there to build documentation.

Reviewer: ```suggestion
        * description is not required by Ansible itself but antsibull-docs does expect it to be there to build documentation.
```

    * This should be ``main`` in the case of an unspecified entry point.
    * This will be the base name of the tasks file to execute, with no ``.yml`` or ``.yaml`` file extension.

.. list-table:: Role Argument Specification Fields
  :width: 90%
  :widths: auto
  :header-rows: 1

    * - Name
      - Description
      - Type
      - Required

>       - Default

Reviewer: Tables with multiple columns tend to be a problem on narrow-width devices, since you have to scroll horizontally to see everything.

    * - short_description
      - A short, one-line description of the entry point.
      - str
      - Yes
      -  
    * - description
      - A longer description that may contain multiple lines.

>       - str

Reviewer: antsibull-docs only accepts lists of strings here, same as for modules and plugins. Since ansible-doc also accepts lists of strings I would at least mention that these are supported as well, or even just mention lists of strings here.

      - No (see notes)
      - null
    * - author
      - Name of the entry point authors.
      - list\n elements: str
      - No
      - []
    * - options
      - Options are often called "parameters" or "arguments".
      - dict (see table below)
      - No
      - {}

Options themselves have a defined structure:
.. list-table:: Role Argument Specification options
  :width: 90%
  :widths: auto
  :header-rows: 1

    * - Name
      - Description
      - Type
      - Required
      - Default

>     * - name

Reviewer: This is a bit confusing, as there is no field called `name` to my knowledge. The option's name is the key of the option dict.

      - The name of the option/argument.
      - str
      - Yes
      -   
    * - description
      - Detailed explanation of what this option does. It should be written in full sentences.

>       -  str

Reviewer: This can also be lists of strings (same as for modules and plugins).

      - No
      - null
    * - type
      - The data type of the option. See :ref:`Argument spec <argument_spec>` for allowed values for ``type``.
      - No
      - str
    * - required
      - Denote if the option is required or not
      - bool
      - No

>       - False

Reviewer: As this is a YAML boolean:
```suggestion
      - false
```

    * - default
      - If an option is not required, the value is set to this if not otherwise provided
      - any (either null or should match Type above)
      - No
      - null
    * - choices
      - List of option values.
      - list
      - No
      - []
    * - elements
      - Specifies the data type for list elements when type is ``list``.
      - str
      - No
      - null
    * - options
      - If this option takes a dict or list of dicts, you can define the structure here.
      - dict (recursively use this same table)
      - No
      - null

.. note::
        * The ``short_description`` is displayed by ``ansible-doc -t role -l``.
        * description is not required by Ansible itself but antsibull-docs does expect it to be there to build documentation.
        * If a type is ``list``, ``elements`` should be specified.
        * Ensure that the default value in the docs matches the default value in the code. The actual
          default for the role variable will always come from ``defaults/main.yml``.

>         * The default field must not be listed as part of the description, unless it requires additional information or conditions.

Reviewer: I would say 'should' and not 'must', since there is no technical reason why this must not be done.
```suggestion
        * The default field should not be listed as part of the description, unless it requires additional information or conditions.
```

>         * If the option's type is boolean, you should use `true/false` for defaults if you want to be compatible with `ansible-lint`.

Reviewer: ```suggestion
        * If the option's type is boolean, you should use ``true``/``false`` for defaults if you want to be compatible with `ansible-lint`.
```

  :width: 90%
  :widths: auto
  :header-rows: 1

  * - Name
    - Description
    - Type
    - Required
    - Default
  * - short_description
    - A short, one-line description of the entry point.
    - str
    - Yes
    -  
  * - description
    - A longer description that may contain multiple lines.
    - list\n elements: str
    - No (see notes)
    - null
  * - author
    - Name of the entry point authors.
    - list\n elements: str
    - No
    - []
  * - options
    - Options are often called "parameters" or "arguments".
    - dict (see table below)
    - No
    - {}

Options themselves have a defined structure

.. list-table:: Role Argument Specification options
  :width: 90%
  :widths: auto
  :header-rows: 1

  * - <Name> (the option itself)
    - Description
    - Type
    - Required
    - Default
  * - name
    - The name of the option/argument.
    - str
    - Yes
    -   
  * - description
    - Detailed explanation of what this option does. It should be written in full sentences.
    - list\n elements: str
    - No
    - null
  * - type
    - The data type of the option. See :ref:`Argument spec <argument_spec>` for allowed values for ``type``.
    - No

>     - str

Reviewer 4: ```suggestion
    - str
    - 
```

Reviewer 4: @bcoca It looks like this one row only has 4 columns instead of 5.

Author: thankx, kept reviewing and could not see this
